### ISSUE TITLE
Make NVDA announce shortcut letters by letter name, not as though in a sentence

### ISSUE THREAD

**Issue by britechguy** [nvaccess/nvda#15301]
### Is your feature request related to a problem? Please describe.
When NVDA is reading the shortcut letter for a menu item, and specifically the letter A in the example I'll give, it is most often read as though the sentence, "There is a dog outside," with a short A, is read.  That makes it very difficult or impossible to pick out the A as the shortcut letter.
In the Thunderbird 102 context menu for a folder tree item, when the following is read, "Favorite Folder  not checked  a  1 of 2," the A that comes between "checked" and "1 of 2" is almost impossible to catch because it's announced as a very quick, short A sound rather than the letter name, A.
### Describe the solution you'd like
Shortcut letters, since it's known that this is what is being announced in the context of menu reading, should always be rendered as the letter name in the language being used to announce it.

### Describe alternatives you've considered
None

### Additional context
None

**Comment by Brian1Gaff** [comment]
OK, are we sure this is NVDA and not the synth? I cannot say I've noticed it in Espeak, but it is harder if using some sappy 5 synths. I guess the proof in the pudding might be to install the sappi5 version of Espeak and set up the voice the same. There are differences here, and its hard to work out what they are.
 Brian

-- 
***@***.*** 
Sent via blueyonder.(Virgin media) 
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.

**Comment by britechguy** [comment]
@XLTechie 

Thank you.  I agree, that no matter what the root cause or causes, the only "central fix point" for this is the screen reader itself.  As I noted in the topic on the NVDA group where I made others aware of this Feature Request ([Make NVDA announce shortcut letters by letter name, not as though in a sentence](https://nvda.groups.io/g/nvda/topic/make_nvda_announce_shortcut/100768978)), what's announced by NVDA is definitely a phrase in 4 parts:  Function Name, Checked/Not Checked State (where applicable), Shortcut Letter, sequence in option list, it should be possible to treat that A, in context, as a standalone item.  The issue comes from the entire phrase being constructed and passed, as a sentence, to the synth or synths that make the A short when so constructed.

I don't think it would hurt for the entire phrase to be "more stilted" overall, so that each element is very much separated from the one preceding it when being announced.  I really don't like the "as a sentence" presentation overall, as it obscures the actual parts even ignoring the short versus long A issue.

Something like:  Favorite Folder.  Unchecked.   A.  1 of 2.
handled as four distinct phrases, gives the correct pause emphasis on the elements it's announcing compared to, "Favorite Folder  not checked  a  1 of 2"

This is, in its own odd way, another variant of the "how human would read to another human were they trying to convey this information" sort of thing.  No one looking at a menu, reading the menu item name, checked/unchecked state (where applicable), shortcut letter, and sequence information would ever, ever say those as though they were a single sentence.  It would still be "quick and dirty" but with pauses so that each element in a spoken list of elements like that is distinct.  And it is critical that any shortcut letter be clearly announced by its letter name, in whatever language it's being announced in, not as though it's a single letter word in continuous speech.

**Comment by CyrilleB79** [comment]
This issue is similar to #10807 and was fixed with #14900 in NVDA 2023.2beta1.

Unfortunately the fix had to be reverted in NVDA 2023.2beta2 due to an issue in CodeFactory's synth (see #15237). In order not to impact Code Factory's synth users, it has been decided to introduce this fix in the next API compatibility breaking release, that is NVDA 2024.1.

You can check this fix testing with [NVDA 2023.2beta1](https://www.nvaccess.org/files/nvda/releases/2023.2beta1/nvda_2023.2beta1.exe) (but not 2023.2beta2 or later).

**Comment by britechguy** [comment]
@CyrilleB79 

While I appreciate the update, and I do, the issue is not fixed until it's fixed, permanently.  I see no use at all testing with a beta version where it's been determined that it will not progress to release candidate or release.  I understand why that fix is being rolled back, but that means it's not a fix at all, at this time.

I probably won't remember by then, but I'll test when 2024.1 comes out.

**Comment by XLTechie** [comment]
@britechguy The reason to test it, is to determine whether it actually provides 
a fix to the problem you described.

It may, but until you try it, we do not know, as your example (which I would 
also like to have resolved), is not specifically described/tested in the PR, 
although it is generally the same thing.

If your use case is solved by that particular beta, then we know this issue may 
be closed after 2024.1. If not, more work in some other direction is required.

**Comment by britechguy** [comment]
@XLTechie 

I read the statement, "This issue is similar to https://github.com/nvaccess/nvda/issues/10807 and was fixed with https://github.com/nvaccess/nvda/pull/14900 in NVDA 2023.2beta1," as saying that the fix was done, and had been tested by someone else already.

If that's not the case, I will be happy to grab the beta and give it a whirl.  "Was fixed" means, to me, exactly what it says.  If more testing is needed, I'm happy to help out.

**Comment by XLTechie** [comment]
All I'm saying, is that from reading the description of the PR, and the phrase 
"is similar to" in what Cyrille said, I would test it before assuming that this 
fix that wasn't a fix actually fixed it.

To repeat: your specific example was not tested directly.
If @CyrilleB79, the author of the fix, says only that your issue is "similar", I 
would want a concrete test before saying that this will definitely be the 
eventual solution.

**Comment by britechguy** [comment]
Since I will be working with the same client this afternoon that served as the trigger for my submitting this feature request, I'll carry along 2023.2beta1 as a portable and see what we get.

That being said, I will repeat, "was fixed" carries exactly that meaning.  When someone tells me "was fixed" that means testing is done and complete before those words are ever uttered.  That's different than "we believe we've got a fix."

I don't typically test, or offer to test, things identified as fixed.  And anyone who knows me from here or elsewhere knows that I'm very happy to test almost whenever requested.  Things identified as fixed don't need testing.  If they do, then you don't know if they're fixed for certain yet, or not.

**Comment by britechguy** [comment]
Whatever it is that was supposed to fix this issue in NVDA 2023.2_beta1 does not do so.

The issue with Thunderbird 102 remains as originally described.  My suspicion, specific to the shortcut letter A being announced as a "short a" as in the sentence, "There's a dog outside," is due to its position in the string being passed to the synthesizer to read.  It makes perfect sense that a phrase such as, "Favorite Folder not checked a 1 of 2," if not handed to the synth as 4 separate pieces, "Favorite Folder," "Not Checked," "A," "1 of 2," would be announced as though that A is an English article ahead of a consonant sound (w).

Contrast this with how NVDA announces two of its own items, "Manage add-ons . . .," and, "Check for add-on updates . . ." in its own Tools menu.  Both of those shortcuts use the letter A, but in both cases that letter is the last thing announced, alone, so there is no way it could be interpreted in context as an article in continuous English speech.

I very much doubt this problem can be solved for the letter A unless it is consistently separated out, as an item all its own, and handed to the synth that way if you want to ensure letter-name announcement for it.

**Comment by Adriani90** [comment]
Somehow related to #49.

**Comment by CyrilleB79** [comment]
> Somehow related to #49.

#49 is about the position where shortcut key is reported; this issue is about how shortcut are reported/spoken. This is quite different.

**Comment by CyrilleB79** [comment]
@britechguy and @XLTechie, just to confirm, is the issue still present in NVDA 2024.4.1?

When testing, please ensure that "Use spelling functionality if supported" is checked (default  value) in speech settings.

**Comment by XLTechie** [comment]
My Notepad example is no longer valid in Windows 11, because of Notepad changes.

Using Brian's Thunderbird example:

OneCore: Still an issue.
E-Speak: Can no longer repro.
MSSAPI 5: Can no longer repro.

In neither case does spelling functionality matter.

**Comment by CyrilleB79** [comment]
OK. Found that in OneCore driver, character mode command is disabled:
```
	def convertCharacterModeCommand(self, command):
		# OneCore's character speech sounds weird and doesn't support pitch alteration.
		# Therefore, we don't use it.
		return None

```
This comes back from initial OneCore implementation (PR #7110 by @jcsteh )

I have tried to re-enable character mode command in OneCore driver. I cannot hear the weird soundin; @jcsteh if you remember what it was, could you be more specific? But I can clearly reproduce the lack of pitch variation, i.e. the pitch is not raised when spelling uppercase letters.

Also, @XLTechie or @britechguy just to confirm to be able to test this without Thunderbird installed:
I have checked "Guess object position information when unavailable" in object presentation settings and go to NVDA menu -> Tools -> Add-on store.
Could you confirm that the issue is reproductible this way?

And what if you just navigate character by character (with left/right arrow) in "AaBbCc"?

**Comment by jcsteh** [comment]
At least for English OneCore voices, some of the character pronunciations are rather odd. They sound quite different to the main voice and I vaguely recall that some of the pronunciations were wrong; maybe "b" or "c" or similar with David, but I can't quite remember. But the bigger concern really is that pitch changes aren't supported.
